demos/guide/quic-server-ssl-poll.c

```diff
@@ -0,0 +1,1413 @@
+        pm->pm_poll_set = new_poll_set;
+        pm->pm_poll_set_sz += POLL_GROW;
+
+    } else if ((pe_num + POLL_DOWNSIZ) < pm->pm_poll_set_sz) {
+        /*
+         * shrink poll set by POLL_DOWNSIZ
+         */
+        new_sz = sizeof (struct poll_event) *
+                 (pm->pm_poll_set_sz - POLL_DOWNSIZ);
+        new_poll_set = (struct poll_event *)OPENSSL_realloc(pm->pm_poll_set,
+                                                            new_sz);
+        if (new_poll_set == NULL)
+            return -1;
+        pm->pm_poll_set = new_poll_set;
+        pm->pm_poll_set_sz -= POLL_GROW;
+    }
+
+    i = 0;
+    DPRINTF(stderr, "%s there %zu events to poll\n", __func__,
+            ossl_list_pe_num(&pm->pm_head));
+    OSSL_LIST_FOREACH(pe, pe, &pm->pm_head) {
+        pe->pe_poll_item.events = pe->pe_want_events;
+        pm->pm_poll_set[i++] = *pe;
+        DPRINTF(stderr, "\t%p (%s) " POLL_FMT " (disabled: " POLL_FMT ")\n",
+                pe, pe_type_to_name(pe),
+                POLL_PRINTA(pe->pe_poll_item.events),
+                POLL_PRINTA(~pe->pe_want_mask));
+    }
+    pm->pm_event_count = i;
+    pm->pm_need_rebuild = 0;
+
+    return 0;
+}
+
+static void destroy_poll_manager(struct poll_manager *pm)
+{
+    struct poll_event *pe, *pe_safe;
+    if (pm == NULL)
+        return;
+
+    OSSL_LIST_FOREACH_DELSAFE(pe, pe_safe, pe, &pm->pm_head) {
+        SSL_free(get_ssl_from_pe(pe));
+	/* ?todo? custom destroy callback on pe */
+        OPENSSL_free(pe);
+    }
+
+    OPENSSL_free(pm->pm_poll_set);
+    OPENSSL_free(pm);
+}
+
+static int pe_accept_qconn(struct poll_event *listener_pe)
+{
+    SSL *listener;
+    SSL *qconn;
+    struct poll_event *qc_pe;
+
+    listener = get_ssl_from_pe(listener_pe);
+    qconn = SSL_accept_connection(listener, 0);
+    if (qconn == NULL)
+        return -1;
+
+    qc_pe = new_qconn_pe(qconn);
+    if (qc_pe != NULL) {
+        if (listener_pe->pe_my_pm->pm_qconn_in(qc_pe) == -1)
+            destroy_pe(qc_pe); /* qc_pe owns qconn */
+        else
+            add_pe_to_pm(listener_pe->pe_my_pm, qc_pe);
+
+    } else {
+        SSL_free(qconn);
+        return -1;
+    }
+
+    return 0;
+}
+
+/*
+ * accept stream from remote peer
+ */
+static int pe_accept_qstream(struct poll_event *qconn_pe)
+{
+    SSL *qconn;
+    SSL *qs;
+    struct poll_event *qs_pe;
+
+    assert(qconn_pe->pe_poll_item.revents & SSL_POLL_EVENT_IS);
+
+    qconn = get_ssl_from_pe(qconn_pe);
+    qs = SSL_accept_stream(qconn, 0);
+    if (qs == NULL)
+        return -1;
+
+    qs_pe = new_pe(qs);
+    if (qs_pe != NULL) {
+        qs_pe->pe_cb_error = pe_handle_qstream_error;
+        qs_pe->pe_cb_in = pe_read_qstream;
+        qs_pe->pe_want_events = SSL_POLL_EVENT_ER;
+
+        if (qconn_pe->pe_poll_item.revents & SSL_POLL_EVENT_ISU) {
+            qs_pe->pe_cb_out = pe_return_error;
+            qs_pe->pe_type = PE_STREAM_UNI_IN;
+        } else if (qconn_pe->pe_poll_item.revents * SSL_POLL_EVENT_ISB) {
```

Comment:
I think you want an & instead of * here, otherwise this check isn't working properly